src/core/mac/sub_mac.cpp

```diff
@@ -328,10 +328,13 @@ void SubMac::ProcessTransmitSecurity(void)
     const ExtAddress *extAddress = nullptr;
     uint8_t           keyIdMode;
 
-    VerifyOrExit(ShouldHandleTransmitSecurity());
     VerifyOrExit(mTransmitFrame.GetSecurityEnabled());
     VerifyOrExit(!mTransmitFrame.IsSecurityProcessed());
 
+    mTransmitFrame.SetKeyId(mKeyId);
```

Comment:
I think this would be incorrect.

If the frame is an indirect retx, then the key id (along with frame counter) is already set by the next layers above `Mac/SubMac` on the frame.   Adding this line here will overwrite the value.

Reply:
Note that indirect frame retransmission can happen much later and on newly prepared frame (when/if the next data poll is received from the child). We need to ensure to use the same frame counter, key id, and seq number for such re-transmissions. 
This is all managed by `DataPollHandler`:
- If a an indirect tx fails, it saves the frame counter, key id and seq number in the `Child` entry (in child table) (`DataPollHandler::HandleSentFrame()`).
- For retx, it sets the same values on the prepared frame `DataPollHandler::HandleFrameRequest()` and mark the frame as a retx (`mIsRetx` set to `true`).

Reply:
> If the frame is an indirect retx, then the key id (along with frame counter) is already set by the next layers

So perhaps higher layers shall set key id regardless security being handled by Radio Platform. This way Radio Platform would know which key to use for this particular frame. What do you think?

Reply:
I think it may be fine to set the key id field in the frame from `SubMac` but we definitely need the `if (!IsARetransmission())` check to ensure to not overwrite the value for indirect retx.

@brianljt, @jwhui  what do you think?